lib/Optimizer/Transforms/ConstPropComplex.cpp

```diff
@@ -0,0 +1,210 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Optimizer/Builder/Intrinsics.h"
+#include "cudaq/Optimizer/Dialect/CC/CCOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "mlir/Dialect/Complex/IR/Complex.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_CONSTPROPCOMPLEX
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+#define DEBUG_TYPE "const-prop-complex"
+
+using namespace mlir;
+
+namespace {
+
+// Replace array ptr casts that throw away the size by a cast to element
+// pointer.
+//
+//%1 = cc.cast %0 : (!cc.ptr<!cc.array<complex<f32> x 4>>) ->
+//! cc.ptr<!cc.array<complex<f32> x ?>>
+// ->
+//%1 = cc.cast %0 : (!cc.ptr<!cc.array<complex<f32> x 4>>) ->
+//! cc.ptr<complex<f32>>
+class CastArrayPtrPattern : public OpRewritePattern<cudaq::cc::CastOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(cudaq::cc::CastOp cast,
+                                PatternRewriter &rewriter) const override {
+
+    auto fromTy = cast.getOperand().getType();
+    auto toTy = cast.getType();
+
+    if (auto ptrFromTy = dyn_cast<cudaq::cc::PointerType>(fromTy)) {
+      if (auto arrayFromTy =
+              dyn_cast<cudaq::cc::ArrayType>(ptrFromTy.getElementType())) {
+        if (auto ptrToTy = dyn_cast<cudaq::cc::PointerType>(toTy)) {
+          if (auto arrayToTy =
+                  dyn_cast<cudaq::cc::ArrayType>(ptrToTy.getElementType())) {
+            if (arrayFromTy.getElementType() == arrayToTy.getElementType()) {
+              auto eleTy = arrayFromTy.getElementType();
```

Comment:
The type `cc.ptr<cc.array<T x ?>>` and `cc.ptr<T>` are not the same and not interchangeable. This rewrite may possibly fail as a consequence.

Reply:
I can try just preserving the original type (ie removing the cast that removes array size) and see if it works.

Reply:
Btw we the IR for constant arrays (without the new const prop pass) uses them interchangeably it seems:

```
// -----// IR Dump Before LiftArrayAlloc (lift-array-value) ('builtin.module' operation) //----- //
module attributes {llvm.data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", llvm.triple = "x86_64-unknown-linux-gnu", quake.mangled_name_map = {__nvqpp__mlirgen__function_test_complex_array_param._Z24test_complex_array_paramSt6vectorISt7complexIfESaIS1_EE = "_Z24test_complex_array_paramSt6vectorISt7complexIfESaIS1_EE", __nvqpp__mlirgen__function_test_complex_constant_array._Z27test_complex_constant_arrayv = "_Z27test_complex_constant_arrayv", __nvqpp__mlirgen__function_test_complex_constant_array2._Z28test_complex_constant_array2v = "_Z28test_complex_constant_array2v", __nvqpp__mlirgen__function_test_complex_constant_array3._Z28test_complex_constant_array3v = "_Z28test_complex_constant_array3v", __nvqpp__mlirgen__function_test_real_array_param._Z21test_real_array_paramSt6vectorIfSaIfEE = "_Z21test_real_array_paramSt6vectorIfSaIfEE", __nvqpp__mlirgen__function_test_real_constant_array._Z24test_real_constant_arrayv = "_Z24test_real_constant_arrayv"}} {
  func.func @__nvqpp__mlirgen__function_test_real_constant_array._Z24test_real_constant_arrayv() attributes {"cudaq-entrypoint", "cudaq-kernel", no_this} {
    %cst = arith.constant 0.70710678118654757 : f64
    %cst_0 = arith.constant 0.000000e+00 : f64
    %0 = cc.alloca !cc.array<f64 x 4>
    %1 = cc.cast %0 : (!cc.ptr<!cc.array<f64 x 4>>) -> !cc.ptr<!cc.array<f64 x ?>>
    %2 = cc.cast %0 : (!cc.ptr<!cc.array<f64 x 4>>) -> !cc.ptr<f64>
    cc.store %cst, %2 : !cc.ptr<f64>
    %3 = cc.compute_ptr %0[1] : (!cc.ptr<!cc.array<f64 x 4>>) -> !cc.ptr<f64>
    cc.store %cst, %3 : !cc.ptr<f64>
    %4 = cc.compute_ptr %0[2] : (!cc.ptr<!cc.array<f64 x 4>>) -> !cc.ptr<f64>
    cc.store %cst_0, %4 : !cc.ptr<f64>
    %5 = cc.compute_ptr %0[3] : (!cc.ptr<!cc.array<f64 x 4>>) -> !cc.ptr<f64>
    cc.store %cst_0, %5 : !cc.ptr<f64>
    %6 = quake.alloca !quake.veq<2>
    %7 = quake.init_state %6, %1 : (!quake.veq<2>, !cc.ptr<!cc.array<f64 x ?>>) -> !quake.veq<2>
    return
  }```

Reply:
I removed the pattern and a added the case handling this cast directly in LifArrayAlloc